modulus/models/vfgn/graph_network.py

```diff
@@ -0,0 +1,562 @@
+        edge_attr_new = edge_attr + edge_attr_res
+
+        return x_new, edge_attr_new, receivers, senders
+
+
+class DecoderNet(torch.nn.Module):
+    def __init__(self, mlp_hidden_size, mlp_num_hidden_layers, output_size):
+        super(DecoderNet, self).__init__()
+        # self.node_mlp = MLPNet(mlp_hidden_size, mlp_num_hidden_layers, output_size)
+        self.mlp = MLPNet(mlp_hidden_size, mlp_num_hidden_layers, output_size, layer_norm=False)
+
+    def forward(self, x):
+        # number of layer is important, or the network will overfit
+        x = self.mlp(x)
+        return x
+
+
+class EncodeProcessDecode(torch.nn.Module):
+    def __init__(self, latent_size, mlp_hidden_size, mlp_num_hidden_layers, num_message_passing_steps, output_size,
+                 device_list=None):
+        super(EncodeProcessDecode, self).__init__()
+        if device_list is None:
+            self.device_list = ['cpu']
+        else:
+            self.device_list = device_list
+
+        self._encoder_network = EncoderNet(mlp_hidden_size, mlp_num_hidden_layers, latent_size)
+
+        self._processor_networks = []
+        for _ in range(num_message_passing_steps):
+            self._processor_networks.append(InteractionNet(mlp_hidden_size, mlp_num_hidden_layers, latent_size))
+        self._processor_networks = torch.nn.ModuleList(self._processor_networks)
+
+        self._decoder_network = DecoderNet(mlp_hidden_size, mlp_num_hidden_layers, output_size)
+
+    def set_device(self,device_list):
+        self.device_list=device_list
+
+    def forward(self, x, edge_attr, receivers, senders):
+        x, edge_attr = self._encoder_network(x, edge_attr)
+        print("Check shape of EncodeProcessDecode, x/ edge_attr: ", x.shape, edge_attr.shape)
+
+        pre_x = x
+        pre_edge_attr = edge_attr
+
+        n_steps = len(self._processor_networks)
+        n_inter = int(n_steps / len(self.device_list))
+
+        i = 0
+        j = 0
+
+        origin_device = x.device
+
+        for processor_network_k in self._processor_networks:
+            p_device = self.device_list[j]
+            processor_network_k = processor_network_k.to(p_device)
+            pre_x = pre_x.to(p_device)
+            pre_edge_attr = pre_edge_attr.to(p_device)
+            receivers = receivers.to(p_device)
+            senders = senders.to(p_device)
+
+            # x, edge_attr, receivers, senders = processor_network_k(x, edge_attr, receivers, senders)
+            diff_x, diff_edge_attr, receivers, senders = checkpoint(processor_network_k, pre_x, pre_edge_attr,
+                                                                    receivers, senders)
+
+            pre_x = x.to(p_device) + diff_x
+            pre_edge_attr = edge_attr.to(p_device) + diff_edge_attr
+            i += 1
+            if i % n_inter == 0:
+                j += 1
+
+        x = self._decoder_network(pre_x.to(origin_device))
+
+        return x
+
+
+class LearnedSimulator(torch.nn.Module):
```

Comment:
I suggest we move the data loading/processing modules to the `datapipes` folder to better organize the code.